experimental/overhead_matching/swag/filter/particle_filter_test.py

```diff
@@ -0,0 +1,243 @@
+
+        new_particles = wag_motion_model(
+            particles, motion_delta, noise_percent, self.generator
+        )
+
+        # Check shape
+        self.assertEqual(new_particles.shape, particles.shape)
+
+        # Check that particles have moved in the x direction
+        self.assertTrue(torch.all(new_particles[:, 0] != 0))
+
+        # Check that noise is applied
+        std_dev = torch.std(new_particles[:, 0])
+        self.assertTrue(std_dev > 0)
+
+    def test_wag_observation_liklihood_from_similarity_matrix(self):
+        similarity_matrix = torch.tensor([[0.8, 0.3], [0.5, 0.9]])
+        sigma = 0.1
+
+        liklihood = wag_observation_liklihood_from_similarity_matrix(
+            similarity_matrix, sigma
+        )
+
+        # Check shape
+        self.assertEqual(liklihood.shape, similarity_matrix.shape)
+
+        # Check that maximum similarity gives max likelihood
+        max_sim_idx = torch.argmax(similarity_matrix)
+        max_lik_idx = torch.argmax(liklihood)
+        self.assertEqual(max_sim_idx, max_lik_idx)
+
+        # Check that values are positive
+        self.assertTrue(torch.all(liklihood > 0))
+
+    def test_wag_update_particle_weights(self):
+        observation_liklihood_matrix = torch.tensor([[0.1, 0.2], [0.3, 0.4]])
+        patch_positions = torch.zeros((2, 2, 2))
+        # Set patch positions at (0,0), (1,0), (0,1), (1,1)
+        patch_positions[0, 0] = torch.tensor([0.0, 0.0])
+        patch_positions[0, 1] = torch.tensor([1.0, 0.0])
+        patch_positions[1, 0] = torch.tensor([0.0, 1.0])
+        patch_positions[1, 1] = torch.tensor([1.0, 1.0])
+
+        # Particles at various positions
+        particles = torch.tensor([
+            [0.1, 0.1],  # Close to (0,0)
+            [0.9, 0.1],  # Close to (1,0)
+            [0.1, 0.9],  # Close to (0,1)
+            [0.9, 0.9]   # Close to (1,1)
+        ])
+
+        weights = wag_update_particle_weights(
+            observation_liklihood_matrix, patch_positions, particles
+        )
+
+        # Check shape
+        self.assertEqual(weights.shape, (particles.shape[0],))
+
+        # Check that weights sum to 1
+        self.assertTrue(torch.isclose(torch.sum(weights), torch.tensor(1.0)))
+
+        # Check that all weights are positive
+        self.assertTrue(torch.all(weights > 0))
+
+    def test_wag_multinomial_resampling(self):
+        particles = torch.tensor([
+            [0.0, 0.0],
+            [1.0, 0.0],
+            [0.0, 1.0],
+            [1.0, 1.0]
+        ])
+
+        # Heavily weight one particle
+        weights = torch.tensor([0.7, 0.1, 0.1, 0.1])
+
+        resampled = wag_multinomial_resampling(
+            particles, weights, self.generator
+        )
+
+        # Check shape
+        self.assertEqual(resampled.shape, particles.shape)
+
+        # With high probability, the first particle should appear multiple times
+        # due to its high weight, but we can't test this deterministically
```

Comment:
Consider adding a `num_samples` argument to the `wag_multinomial_resampling()`  function. Sample 1000 particles and it should be safe to assume that you have the most number of the first.